# Coding Temple's Data Analytics Course
---
## Python II: Numpy & CSV's

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>


6) <b>Numpy</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Python List Comparison <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) In-Class Exercise #6 <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Importing <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Creating an NDArray <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - np.array() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - np.zeros() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - np.ones() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - np.arange() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Making Lists into NDArrays <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Performing Calculations on NDArrays <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Summation <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Difference <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Multiplication <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Division <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Standard Distribution <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Mean <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Median <br> 
 &nbsp;&nbsp;&nbsp;&nbsp; f) Numpy Subsetting <br>
 &nbsp;&nbsp;&nbsp;&nbsp; g) Multi-dimensional Arrays <br>
 &nbsp;&nbsp;&nbsp;&nbsp; h) Indexing NDArrays <br>
 &nbsp;&nbsp;&nbsp;&nbsp; i) Checking NDArray Type <br>
 &nbsp;&nbsp;&nbsp;&nbsp; j) Altering NDArray Type <br>
 &nbsp;&nbsp;&nbsp;&nbsp; k) Checking the Shape <br>
 &nbsp;&nbsp;&nbsp;&nbsp; l) Altering the Shape <br>
 &nbsp;&nbsp;&nbsp;&nbsp; m) In-Class Exercise #7 <br>
 &nbsp;&nbsp;&nbsp;&nbsp; n) Complex Indexing & Assigning <br>
 &nbsp;&nbsp;&nbsp;&nbsp; o) Elementwise Multplication <br>
 &nbsp;&nbsp;&nbsp;&nbsp; p) np.where() <br>
 &nbsp;&nbsp;&nbsp;&nbsp; q) Random Sampling <br>

7) <b>Working With CSV's</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Imports <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Reading a CSV <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Loading a CSV's Data <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Checking Number of Records <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [2]:
# Let's take a look at how we can create a function to add two to a variable:
def add_two(x):
    return x + 2

# Then we would test it:
print(add_two(4))

# This is great when I am looking to reuse a function. But what if I only needed it for a single use app?
# This is where lambda can come into play!
a_var = 4

# lambda >input variable<: whatever we want to do
print((lambda x: x+2)(a_var))

6
6


#### Saving to a Variable

In [4]:
# We can save lambda functions to a variable to make them more reusable:
f_test = lambda x: x+2

f_test(4)

6

In [5]:
f_test(78)

80

#### Multiple Inputs

In [6]:
# What if we had 3 numbers and we wanted to be able to take the product of all those numbers?

# EXACTLY the same as a normal function:
def product(x,y,z):
    return x*y*z


print((lambda x,y,z: z* y * x)(3,5,8))

120


#### Passing a Lambda into a Function

In [7]:
def multiply(function, num):
    """
    function expects to see a function. (Lambda function)
    num expects to see a single integer input
    """
    return function(num)

print(multiply(lambda x: x*x, 4))

# Not only can I do this with lambda, but I can use ordinary functions as well!
def squared_nums(num):
    return num ** 2

print(multiply(squared_nums, 4))

16


#### Returning a Lambda from a Function

In [10]:
def multiply_test(num):
    return num * 4

# Take this function, and run it inside another function:
def return_func():
    test = 4
    def multiply(num):
        return num * 2
    return multiply(test)

print(return_func())

# I can do the same thing using lambda:
def return_lambs(b,c):
    return lambda x,a: x+a+b+c

r_lamb = return_lambs(4,5)
print(r_lamb)
print(r_lamb(5,9))

8
<function return_lambs.<locals>.<lambda> at 0x00000206B0BE3880>
23


#### If Statements within Lambdas

In [12]:
# Lambda functions are set up to be able to take in conditional statements!
# Lambda x: True if (conditional) else False
# Lambda x: do something if (conditional is True) else, do something else
# An example:
f_conditional = lambda num: num * 2 if num > 10 else num + 2

print(f_conditional(8))
print(f_conditional(20))

10
40


In [15]:
lst = [1,2,3,4,5,6,7,8,9,10,11,12,13,14]

# Run the lambda on the list:
[f_conditional(num) for num in lst]

[3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 22, 24, 26, 28]

#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [16]:
# Write out the function:
f = lambda x: x ** 3

# Invoke the function:
f(10)

1000

## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [17]:
numbers = [4,11,20,3,15,20,11,3,2,6,5,1,44,1]

# Normally, in order to use a function on this, we would need to iterate through:
def squared_normal(val):
    return val ** 2
lst = []
for num in numbers:
    lst.append(squared_normal(num))
    
print(lst)

[16, 121, 400, 9, 225, 400, 121, 9, 4, 36, 25, 1, 1936, 1]


In [22]:
'''
map(function, iterable)
An iterable is something we can loop over; such as a list, dict, tuple, set, or a string!
Normally, you would use map with a predefined function, but we can also use it directly with a lambda statement!
'''

def squared_map(num, num1):
    if num < 10 and num1 < 10:
        return num ** 2, num1 **2
    else:
        return num, num1
    
numbers = [4,11,20,3,15,20,11,3,2,6,5,1,44,1]
more_nums = [4,10,3,2,6,10,15,3,4,7,6,0,42,1]

# In order to use the squared_map function without a map statement, we would need to iterate through both lists at once.
# To do so effectively, we will utlize the zip() function, which brings together two iterable objects and 
# Returns the values at each index position, up till the end of the shortest length of the iterables.
# for val1, val2 in zip(numbers, more_nums):
#     print(squared_map(val1, val2))
    
list(map(squared_map, numbers, more_nums))

[(16, 16),
 (11, 10),
 (20, 3),
 (9, 4),
 (15, 6),
 (20, 10),
 (11, 15),
 (9, 9),
 (4, 16),
 (36, 49),
 (25, 36),
 (1, 0),
 (44, 42),
 (1, 1)]

#### Using Lambda's with Map

In [24]:
"""
Instead of having to build out a function that we plan on using one time,
We can use a lambda inside the map function itself instead. This allows us
to do everything we need to do in a single line of code

Syntax for this would look like:
map(lambda x: do something, iterable)
"""
#Remember to always use the list function with the map function!
# Run Lambda over two lists, based on conditional
list(map(lambda x,y: (x **2, y **2) if x < 10 and y < 10 else (x,y), numbers, more_nums))

[(16, 16),
 (11, 10),
 (20, 3),
 (9, 4),
 (15, 6),
 (20, 10),
 (11, 15),
 (9, 9),
 (4, 16),
 (36, 49),
 (25, 36),
 (1, 0),
 (44, 42),
 (1, 1)]

#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [25]:
list(map(lambda x: x * 2 - 1, numbers))

[7, 21, 39, 5, 29, 39, 21, 5, 3, 11, 9, 1, 87, 1]

## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [28]:
# Syntax for this looks like:
# list(filter(function, iterable(s)))
# A lot like a mask
names = ['Bob', 'Andy', 'Max','Evan', 'Angela']

def a_names(name):
    # In this function, we will check to see if the first letter of each name begins with an 'A' 
    if name[0].lower() == 'a':
        return True
    else:
        return False

# Our function here returns a boolean as a value, just like any other conditional filter.
# We can use this inside a filter to filter out our information with ease
list(filter(a_names, names))

['Andy', 'Angela']

#### Using Lambda's with Filter()

In [30]:
# We can also use lambdas!
# This is great if you are trying to filter data with a single-use function instead of having to define one

# When we use a lambda, or comprehension, we will typically write out the value first, then the conditional
# The best way to think about this is like a reversed for-loop
list(filter(lambda name: True if name[0].lower() == 'a' else False, names))

# List comprehension:
[True if name[0].lower() == 'a' else False for name in names]

['Andy', 'Angela']

#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [33]:
import statistics
statistics.mean(numbers)

# Number 1: Using vanilla Python:
print(list(filter(lambda x: True if x >= sum(numbers) / len(numbers) else False, numbers)))

# Number 2: Using the statistics module:
print(list(filter(lambda x: True if x >= statistics.mean(numbers) else False, numbers)))

[11, 20, 15, 20, 11, 44]
[11, 20, 15, 20, 11, 44]


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [ ]:
# Start off with defining the function:
def add_nums_breakable(num):
    # Without a base case, we will end up infinitely looping through our program
    if num >= 20000:
        return num + add_nums_breakable(num -1)
    else:
        return num

In [34]:
def add_nums(num):
    if num <=1:
        print('add_nums(1) = 1')
        return num
    else:
        print(f'add_nums({num}) = {num} + add_nums({num - 1})')
        return num + add_nums(num-1)

add_nums(5)

add_nums(5) = 5 + add_nums(4)
add_nums(4) = 4 + add_nums(3)
add_nums(3) = 3 + add_nums(2)
add_nums(2) = 2 + add_nums(1)
add_nums(1) = 1


15

#### Writing a Factorial Function

In [35]:
# 5! = 5 * 4 * 3 * 2 * 1
def factorial(num):
    # Base Case!
    if num <=1:
        print('factorial(1) = (1)')
        return num
    else:
        print(f'factorial({num}) = {num} * factorial({num - 1})')
        return num * factorial(num -1)
    
print(factorial(5))

factorial(5) = 5 * factorial(4)
factorial(4) = 4 * factorial(3)
factorial(3) = 3 * factorial(2)
factorial(2) = 2 * factorial(1)
factorial(1) = (1)
120


#### In-Class Exercise #4 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [36]:
def subtract(num1, num2):
    if num2 == 0:
        return num1
    else:
        print(num1, num2)
        return subtract(num1-1, num2 - 1)

subtract(25,5)

25 5
24 4
23 3
22 2
21 1


20

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [42]:
[i for i in range(10)]

def my_range(stop, start=0, step=1):
    while start < stop:
        yield start
        start += step
        
        
        
print([i for i in my_range(10)])

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]


#### In-Class Exercise #5 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [44]:
def squared_nums(num):
    while num > 0:
        yield num ** 2
        num -= 1
        
[i for i in squared_nums(5)]

[25, 16, 9, 4, 1]

## Numpy <br>

<p>NumPy is a library for the Python programming language, adding support for large, multi-dimensional arrays and matrices, along with a large collection of high-level mathematical functions to operate on these arrays.</p>
<ul>
    <li>Shape = Rows & Columns</li>
    <li>Matrix = Entire Array</li>
    <li>Vector = Variables to be applied (same vector as the one used in physics)</li>
    <li>Array = Similar to lists</li>
</ul>

#### Python List Comparison

Lists are flexible, dynamic python objects that do their job quite well. But they do not support some mathematical operations in an intuitive way. Consider the summation of two lists, $l_1$ and $l_2$

In [46]:
l1 = [1,2,3]
l2 = [4,5,6]

print(f'Sum of the lists: {l1+l2}')

# Apply a subtraction element-wise to the list:
print(f'Try to subtract like this: {l1-l2}')

Sum of the lists: [1, 2, 3, 4, 5, 6]


TypeError: unsupported operand type(s) for -: 'list' and 'list'

<p>If we wanted to sum lists elementwise, we could write our own function that does the job entirely within the framework of python</p>

#### In-Class Exercise #6 - Write a function that sums the indexes of two lists <br>
<p>Ex: [2, 3, 4] + [1, 5, 2] = [3, 8, 6]</p>

In [48]:
list1 = [5,5,10]
list2 = [2,4,1]

# Output: [7,9,11]
def sum_list(l1,l2):
    final = [] # Create an empty list
    i = 0 # Index counter:
    while i < len(l1):
        final.append(l1[i] + l2[i])
        i += 1
    return final

print(sum_list(list1, list2))

def sum_list_comp(l1, l2):
    return [l1[i] + l2[i] for i in range(len(l1))]

print(sum_list_comp(list1, list2))

[7, 9, 11]
[7, 9, 11]


We would have to write a similar function for all the possible operands that we could consider for list arithmatic. This is time consuming and inefficient. Moreover, once the lists in question become nested, mimicing the behavior of true matrices, the problem gets worse. Complicated indexing is necessary, just to allow for the most basic matrix operations common throughout science and engineering. Imagine writing a matrix multiplication function using python syntax in a general way, such that it returns a matrix-matrix or matrix-vector product:

\begin{align}
(n \times x) \times (x \times m) \rightarrow (n \times m)
\end{align}

\begin{align}
\begin{bmatrix}
c_{0,0} & ... & c_{0,n} \\
\vdots & \ddots & \vdots \\
c_{m,0} & ... & c_{m,n}
\end{bmatrix}
=
\begin{bmatrix}
a_{0,0} & ... & a_{0,x} \\
\vdots & \ddots & \vdots \\
a_{n,0} & ... & a_{n,x}
\end{bmatrix}
\begin{bmatrix}
b_{0,0} & ... & b_{0,m} \\
\vdots & \ddots & \vdots \\
b_{x,0} & ... & b_{x,m}
\end{bmatrix}
\end{align}

Another Example:

<img src="https://www.mathsisfun.com/algebra/images/matrix-multiply-a.svg">

---

The "Dot Product" is where we multiply matching members, then sum up:

(1, 2, 3) • (7, 9, 11) = 1×7 + 2×9 + 3×11
    = 58

We match the 1st members (1 and 7), multiply them, likewise for the 2nd members (2 and 9) and the 3rd members (3 and 11), and finally sum them up.

<hr>

So, let's instantiate a matrix $\mathcal{M}$ and a vector $\vec{v}$ and write a function that does the multiplication ourselves.

In [51]:
def matrix_multiplication(A,B):
    zero_list = [[0 for i in range(len(B[0]))] for i in range(len(A))]
    print(zero_list)
    inner_dimensions = len(A[0])
    print(inner_dimensions)
    n_dimension = len(zero_list)    
    print(n_dimension)
    m_dimension = len(zero_list[0])
    print(m_dimension)
    
    # Now we can loop through our matrix using the dimensions:
    for i in range(n_dimension):
        for j in range(m_dimension):
            element = 0
            for x in range(inner_dimensions):
                element += A[i][x] * B[x][j]
            zero_list[i][j] = element
    return zero_list
print(matrix_multiplication([[0,1,0],[0,2,0],[0,3,0]], [[1],[2],[3]]))

[[0], [0], [0]]
3
3
1
[[2], [4], [6]]


#### Importing

In [52]:
import numpy as np # Aliasing

#### Creating an NDArray <br>
NumPy is based around a class called the $\textit{NDArray}$, which is a flexible vector / matrix class that implements the intuitive matrix and vector arithmatic lacking in basic Python. Let's start by creating some NDArrays:

##### - np.array()

In [53]:
arr_1 = np.array([1,2,3])
print(arr_1)
print(type(arr_1))

[1 2 3]
<class 'numpy.ndarray'>


##### - np.zeros()

In [56]:
# Create a shape variable. It will reference the 2-D shape I want my matrix to have:
shape = (3,3)
# We create the matrix!
M = np.zeros(shape, int)
M

array([[0, 0, 0],
       [0, 0, 0],
       [0, 0, 0]])

##### - np.ones()

In [57]:
arr2 = np.ones(shape, int)
print(arr2)

[[1 1 1]
 [1 1 1]
 [1 1 1]]


##### - np.arange()

In [58]:
np.arange(1,12)

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11])

##### - Making Lists into NDArrays

In [59]:
l3 = [4,5,6]

# We just use the .array() function to turn this list into an array instead:
arr_list = np.array(l3)
arr_list

array([4, 5, 6])

#### Performing Calculations on NDArrays

##### - Summation

In [63]:
arr5 = np.array([5,10,15,20,25])
arr10 = np.array([10,20,30,40,50])
arr5 + arr10

array([15, 30, 45, 60, 75])

##### - Difference

In [61]:
arr10 - arr5

array([ 5, 10, 15, 20, 25])

##### - Multiplication

In [64]:
arr5 * arr10

array([  50,  200,  450,  800, 1250])

##### - Division

In [68]:
(arr10 / arr5).astype(int)


array([2, 2, 2, 2, 2])

##### - Standard Deviation

In [69]:
print(np.std(arr10))

14.142135623730951


##### - Mean

In [71]:
int(np.mean(arr5))

15

##### - Median

In [73]:
int(np.median(arr5))

15

#### Numpy Subsetting

In [74]:
mask = arr10 > 20
arr10[mask]

array([30, 40, 50])

#### Multi-dimensional Arrays <br>
NumPy seamlessly supports multidimensional arrays and matrices of arbitrary dimension without nesting NDArrays. NDArrays themselves are flexible and extensible and may be defined with such dimensions, with a rich API of common functions to facilitate their use. Let's start by building a two dimensional 3x3 matrix by conversion from a nested group of core python lists $M = [l_0, l_1, l_2]$:

In [75]:
a_list = [0,1,2]
b_list = [3,4,5]
c_list = [6,7,8]

# Step 1: Convert the lists into a matrix that we can actually work with:
M = [a_list, b_list, c_list]
print(f'This is a nested list structure: {M}')

# Step 2: We can cast it directly into a matrix now. 
# When I do so, the dimensions of that matrix are inferred by Numpy for us.
M = np.array(M)
print(f'Matrix Structure: {M}')

This is a nested list structure: [[0, 1, 2], [3, 4, 5], [6, 7, 8]]
Matrix Structure: [[0 1 2]
 [3 4 5]
 [6 7 8]]


#### Indexing NDArrays <br>
<p> Similar to lists within lists; however, the syntax looks more like C programming language.... It is [1, 2] to access the second row, third element.</p>

In [77]:
# Because of the two dimensional structure, single level indexing is going to return values in the entire row:
print(f'Get the first array out of the matrix: {M[0]}')

# Multi-level indexing:
print(f'Get the middle value of the array: {M[1][1]}')

Get the first array out of the matrix: [0 1 2]
Get the middle value of the array: 4


#### Assigning Values in NDArrays

In [80]:
# We can assign values to array in a Matrix in two different ways

# Multi-level indexing way
# This separates out the statement into two indicies
M[1][1] = 9
# print(M)

# Another way to go about this is the give a list of indicies. 
# The list has the first index, then the second index.
M[0,1] = 8
print(M)

[[0 8 2]
 [3 9 5]
 [6 7 8]]


[[0 8 2]
 [3 9 5]
 [6 7 8]]


<p>Notice above how we ended up with a 1 in the target element's place. This is a data type issue. The .dtype() method is supported by all NDArrays, as well as the .astype() method for casting between data types:</p>

#### Checking NDArray Type

In [81]:
M.dtype

dtype('int32')

#### Altering NDArray Type

In [83]:
# Let's take a look at the type of the data before we make any changes:
t = M
print(f'Type before changes: {t.dtype}')

# Change the data to be a float instead?
# We will utilize the .astype() function
changed = t.astype(np.float32)
print(changed)
print(f'Type after the change: {changed.dtype}')

Type before changes: int32
[[0. 8. 2.]
 [3. 9. 5.]
 [6. 7. 8.]]
Type after the change: float32


#### Checking the Shape <br>
The behavior and properties of an NDArray are often sensitively dependent on the $\textit{shape}$ of the NDArray itself. The shape of an array can be found by calling the .shape method, which will return a tuple containing the array's dimensions:

In [85]:
# In a 2-D structure, the shape will displayed as: (rows, columns) or (rows x columns)
print(changed.shape)

# 1-D structure, or a vector. When the shape is displayed, the information returned is: (rows,) or (rows). 
# These do not contain an actual shape, but rather a magnitude
print(arr5.shape)

(3, 3)
(5,)


#### Altering the Shape <br>
<p>As long as the number of elements remains fixed, we can reshape NDArrays at will:</p>

In [86]:
# To do so, we can use the .reshape() function from numpy.
# The syntax for this is .reshape((rows,columns))
print(changed.shape)
print(f'Now we will reshape the matrix into a vector: {changed.reshape(9,1).astype(int)}')

(3, 3)
Now we will reshape the matrix into a vector: [[0]
 [8]
 [2]
 [3]
 [9]
 [5]
 [6]
 [7]
 [8]]


In [ ]:
# Keep in mind: When we reshape a matrix, the shape of the numbers do matter
# (9,1) is much different than that of a (1,9)
print(changed.reshape(1,9).astype(int))

#### In-Class Exercise #7 - Create a matrix of range 0 up to 16 and reshape it into a 4x4 matrice

In [ ]:
# IF YOU GET THIS DONE BEFORE TIME IS UP:
# Try to create a matrix of floating point values from 0-16 and reshape that into 4x4

# One way using Python built-ins
x = np.array(range(0, 16)).reshape(4,4)
print(x)

# Another way using Numpy functions
y = np.arange(0, 16).reshape(4,4)
print('\n',y)

#### Complex Indexing & Assinging

In [90]:
# Reset the matrix:
M = np.zeros((6,6)).astype(int)
print('Resetting the Matrix... Please wait:')
print(f'Before assigning values:\n {M}')

# Set every first element in each column to the value of 1:
M[:,0] = 1
print(f'\nAfter assigning 1 to the first column:\n{M}')

# set all elements in the axis 3 to 5 to be equal to the value of 5
M[2:5, :] = 5
print(f'\nAfter assigning 5:\n{M}')

# Reset back to zero!
# This will be an example of the element-wise multiplication of a Matrix object:
M = M * 0
print(f'Reset was confirmed:\n{M}')

# set the 2nd and 3rd columns equal to the value of 2:
M[:, 1:3] = 2
print(f'Assigning 2 to the second and third columns:\n{M}')

Resetting the Matrix... Please wait:
Before assigning values:
 [[0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]]

After assigning 1 to the first column:
[[1 0 0 0 0 0]
 [1 0 0 0 0 0]
 [1 0 0 0 0 0]
 [1 0 0 0 0 0]
 [1 0 0 0 0 0]
 [1 0 0 0 0 0]]

After assigning 5:
[[1 0 0 0 0 0]
 [1 0 0 0 0 0]
 [5 5 5 5 5 5]
 [5 5 5 5 5 5]
 [5 5 5 5 5 5]
 [1 0 0 0 0 0]]
Reset was confirmed:
[[0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]]
Assigning 2 to the second and third columns:
[[0 2 2 0 0 0]
 [0 2 2 0 0 0]
 [0 2 2 0 0 0]
 [0 2 2 0 0 0]
 [0 2 2 0 0 0]
 [0 2 2 0 0 0]]


#### Elementwise Multiplication

<p>As long as the shapes of NDArrays are 'compatible', they can be multiplied elementwise, broadcasted, used in inner products, and much much more. 'Compatible' in this context can mean compatible in the linear algebraic sense, i.e. for inner products and other matrix multiplication, or simply sharing a dimension in such a manner that broadcasting 'makes sense'. Here are some examples of this:</p>

In [96]:
M *= 0 
matrix = M

# Create the vector of range 6:
vector = np.arange(6)
print(matrix)
print('\n')
print(vector)

matrix[0] += vector
print(matrix)

matrix *= vector
print(matrix)

[[0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]]


[0 1 2 3 4 5]
[[0 1 2 3 4 5]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]
 [0 0 0 0 0 0]]
[[ 0  1  4  9 16 25]
 [ 0  0  0  0  0  0]
 [ 0  0  0  0  0  0]
 [ 0  0  0  0  0  0]
 [ 0  0  0  0  0  0]
 [ 0  0  0  0  0  0]]


#### np.where() <br>
<p>If statement within NDArrays that allows you to run conditionals on the entire array</p>

In [99]:
print(f'{np.where(M>2)}')

# Conditional masking
mask = np.where(M==1)
M[mask] = 2
print(M)

(array([0, 0, 0, 0], dtype=int64), array([2, 3, 4, 5], dtype=int64))
[[ 0  2  4  9 16 25]
 [ 0  0  0  0  0  0]
 [ 0  0  0  0  0  0]
 [ 0  0  0  0  0  0]
 [ 0  0  0  0  0  0]
 [ 0  0  0  0  0  0]]


#### Random Sampling <br>
<p>NumPy provides machinery to work with random numbers - something often needed in a broad spectrum of data science applications.</p>

In [129]:
# np.random.uniform(), uniform creates a floating point value
# np.random.randint(), creates a random integer value

# To start, we will set a random seed. A random seed allows us to reproduce our random code. Meaning everytime we run it,
# Instead of getting new random numbers, we will end up with the same numbers.
np.random.seed(42)

# A single call to generate a single random number between 0 and 1
# Remember that this is a floating point value!
print(f'Here is a random float: {np.random.uniform()}')

# We can also pass in some parameters and bounds
print(f'Here is a random float between 0 and 1 million: {np.random.uniform(0, 1e6)}')

# We can generate a multitude of random numbers using uniform!
print(f'Here are three random floats between 0 and 10: {np.random.uniform(0,10,3)}')
print(type(np.random.uniform(0,10,3)))

# What if I wanted to generate a Matrix instead of an array?
print(f'Here is a 3x3 matrix with floats between 0 and 10: \n{np.random.uniform(0,10, (3,3))}')

# What if I needed integer values instead of floats?
print(f'Here is how we can generate random integer values instead!:\n{np.random.randint(0,10, (3,3))}')

0.3745401188473625
950714.3064099161
[7.58794548 6.38792636 2.40416776]
[[3 7 8 5 4 8]
 [8 3 6 5 2 8]
 [6 2 5 1 6 9]
 [1 3 7 4 9 3]
 [5 3 7 5 9 7]
 [2 4 9 2 9 5]]


## Working With CSV's

#### Imports

In [130]:
import csv

#### Reading a CSV

In [131]:
def open_csv(filepath, d = ','):
    # Define a list to store the data in:
    data = []
    
    # Use the with keyword and utf-8-sig encoding to load data from .csv!
    with open(filepath, encoding='utf-8-sig') as my_data:
        info = csv.reader(my_data, delimiter=d)
        for row in info:
            data.append(row)
            
    return data

csv_data = open_csv(r'C:\Users\Alex Lucchesi\coding-temple\coding_temple_data_analytics_ft\week-3\data\redsox_2017_hitting.txt')
csv_data

[['Rk',
  'Pos',
  'Name',
  'Age',
  'G',
  'PA',
  'AB',
  'R',
  'H',
  '2B',
  '3B',
  'HR',
  'RBI',
  'SB',
  'CS',
  'BB',
  'SO',
  'BA',
  'OBP',
  'SLG',
  'OPS',
  'OPS+',
  'TB',
  'GDP',
  'HBP',
  'SH',
  'SF',
  'IBB'],
 ['1',
  'C',
  'Christian Vazquez',
  '26',
  '99',
  '345',
  '324',
  '43',
  '94',
  '18',
  '2',
  '5',
  '32',
  '7',
  '2',
  '17',
  '64',
  '.290',
  '.330',
  '.404',
  '.735',
  '91',
  '131',
  '14',
  '3',
  '0',
  '1',
  '0'],
 ['2',
  '1B',
  'Mitch Moreland',
  '31',
  '149',
  '576',
  '508',
  '73',
  '125',
  '34',
  '0',
  '22',
  '79',
  '0',
  '1',
  '57',
  '120',
  '.246',
  '.326',
  '.443',
  '.769',
  '99',
  '225',
  '14',
  '6',
  '0',
  '5',
  '6'],
 ['3',
  '2B',
  'Dustin Pedroia',
  '33',
  '105',
  '463',
  '406',
  '46',
  '119',
  '19',
  '0',
  '7',
  '62',
  '4',
  '3',
  '49',
  '48',
  '.293',
  '.369',
  '.392',
  '.760',
  '100',
  '159',
  '11',
  '2',
  '2',
  '4',
  '4'],
 ['4',
  'SS',
  'Xander Bogaerts',
  '

#### Loading a CSV's Data 

In [143]:
# using numpy:
FIELDS = ['Rk','Pos','Name','Age','G','PA','AB','R','H','2B','3B','HR','RBI','SB',
          'CS','BB','SO','BA','OBP','SLG','OPS','OPS+','TB','GDP','HBP','SH','SF','IBB']

DATATYPES = [('rk', 'i'), ('pos', '|S25'), ('name', '|S25'), ('age', 'i'), ('g', 'i'), ('pa', 'i'), ('ab', 'i'),
                ('r', 'i'), ('h', 'i'), ('2b', 'i'), ('3b', 'i'), ('hr', 'i'), ('rbi', 'i'), ('sb', 'i'), ('cs', 'i'),
                ('bb', 'i'), ('so', 'i'), ('ba', 'f'), ('obp', 'f'), ('slg', 'f'), ('ops', 'f'), ('opsp', 'i'),
                ('tb', 'i'), ('gdp', 'i'), ('hbp', 'i'), ('sh', 'i'), ('sf', 'i'), ('ibb', 'i')]

# Instead of using the csv.reader like we just did, we will load the data into a numpy array and calculate some results:
def load_data(filepath, d = ','):
    data = np.genfromtxt(filepath, delimiter=d, skip_header=1,
                         usecols = np.arange(0,24), invalid_raise=False,
                         names = FIELDS, dtype=DATATYPES)
    return data

# Load in the data
bs2017 = load_data(r'C:\Users\Alex Lucchesi\coding-temple\coding_temple_data_analytics_ft\week-3\student_notebooks\redsox_2017_hitting.txt')
bs2018 = load_data(r'C:\Users\Alex Lucchesi\coding-temple\coding_temple_data_analytics_ft\week-3\student_notebooks\redsox_2018_hitting.txt')

# Calculate the average batting average
length = len(bs2018['BA'])
print(f'Average of the Batting Average Column is: {round((sum(bs2018["BA"]) / length)*100, 2)}')

# Sum of total hits
print(f'Total Hits: {int(sum(bs2018["H"]))}')

# Home Runs Total:
print(f"Total Home Runs: {int(sum(bs2018['HR']))}")

The Average of the Batting Average column is: 27.84
Total Hits: 1459
Total Home Runs: 168
Total Percentage of hits to home runs: 0.11514736120630568


In [144]:
bs2017

array([( 1, b'C', b'Christian Vazquez', 26,  99, 345, 324,  43,  94, 18, 2,  5,  32,  7, 2, 17,  64, 0.29 , 0.33 , 0.404, 0.735,  91, 131, 14),
       ( 2, b'1B', b'Mitch Moreland', 31, 149, 576, 508,  73, 125, 34, 0, 22,  79,  0, 1, 57, 120, 0.246, 0.326, 0.443, 0.769,  99, 225, 14),
       ( 3, b'2B', b'Dustin Pedroia', 33, 105, 463, 406,  46, 119, 19, 0,  7,  62,  4, 3, 49,  48, 0.293, 0.369, 0.392, 0.76 , 100, 159, 11),
       ( 4, b'SS', b'Xander Bogaerts', 24, 148, 635, 571,  94, 156, 32, 6, 10,  62, 15, 1, 56, 116, 0.273, 0.343, 0.403, 0.746,  95, 230, 17),
       ( 5, b'3B', b'Rafael Devers', 20,  58, 240, 222,  34,  63, 14, 0, 10,  30,  3, 1, 18,  57, 0.284, 0.338, 0.482, 0.819, 111, 107,  5),
       ( 6, b'LF', b'Andrew Benintendi', 22, 151, 658, 573,  84, 155, 26, 1, 20,  90, 20, 5, 70, 112, 0.271, 0.352, 0.424, 0.776, 102, 243, 16),
       ( 7, b'CF', b'Jackie Bradley Jr.', 27, 133, 541, 482,  58, 118, 19, 3, 17,  63,  8, 3, 48, 124, 0.245, 0.323, 0.402, 0.726,  89, 194,  8

#### Summing the top 5 hitters for HR's

In [146]:
# Step 1: Grab all the HR data into a single variable
all_hr = bs2017['HR']

# Step 2: Grab all the player names
all_names = bs2017['Name']

# Step 3: Create a new list for the new names and the amount of home runs
new_names = []

# Step 4: Decode the byte strings of all player names
for name in range(len(all_names)):
    new_names.append(all_names[name].decode('utf-8'))

# Step 5: Grab all the data and place inside of a zipped list:
zipped_values = list(zip(new_names, all_hr))

# Step 6: Sort values based on home run data
sorted_hr = sorted(zipped_values, key= lambda x: x[1], reverse = True)
sorted_hr[:5]

[('Mookie Betts', 24),
 ('Hanley Ramirez', 23),
 ('Mitch Moreland', 22),
 ('Andrew Benintendi', 20),
 ('Jackie Bradley Jr.', 17)]